pkg/volume/util/operationexecutor/operation_generator.go

```diff
@@ -1672,6 +1689,183 @@ func (og *operationGenerator) GenerateExpandVolumeFunc(
+	}
+
+	eventRecorderFunc := func(err *error) {
+		if *err != nil {
+			og.recorder.Eventf(pvc, v1.EventTypeWarning, kevents.VolumeResizeFailed, (*err).Error())
+		}
+	}
+
+	return volumetypes.GeneratedOperations{
+		OperationName:     "expand_volume",
+		OperationFunc:     expandVolumeFunc,
+		EventRecorderFunc: eventRecorderFunc,
+		CompleteFunc:      util.OperationCompleteHook(util.GetFullQualifiedPluginNameForVolume(volumePlugin.GetPluginName(), volumeSpec), "expand_volume"),
+	}, nil
+}
+
+func (og *operationGenerator) expandAndRecoverFunction(resizeOpts inTreeResizeOpts) inTreeResizeResponse {
+	pvc := resizeOpts.pvc
+	pv := resizeOpts.pv
+	resizerName := resizeOpts.resizerName
+	volumePlugin := resizeOpts.volumePlugin
+	volumeSpec := resizeOpts.volumeSpec
+
+	pvcSpecSize := pvc.Spec.Resources.Requests[v1.ResourceStorage]
+	pvcStatusSize := pvc.Status.Capacity[v1.ResourceStorage]
+	pvSize := pv.Spec.Capacity[v1.ResourceStorage]
+
+	resizeResponse := inTreeResizeResponse{
+		pvc:          pvc,
+		pv:           pv,
+		resizeCalled: false,
+	}
+
+	// by default we are expanding to full-fill size requested in pvc.Spec.Resources
+	newSize := pvcSpecSize
+	resizeStatus := v1.PersistentVolumeClaimNoExpansionInProgress
+	if pvc.Status.ResizeStatus != nil {
+		resizeStatus = *pvc.Status.ResizeStatus
+	}
+	var allocatedSize *resource.Quantity
+	t, ok := pvc.Status.AllocatedResources[v1.ResourceStorage]
+	if ok {
+		allocatedSize = &t
+	}
+
+	// pv is not of requested size yet and hence will require expanding
+	if pvSize.Cmp(pvcSpecSize) < 0 {
+		// if we are in any of following states, that means expansion has already began
+		// in which case we should try and finish the expansion to previous size
+		// rather than starting expansion to a new size
+		if resizeStatus == v1.PersistentVolumeClaimControllerExpansionInProgress ||
+			resizeStatus == v1.PersistentVolumeClaimNodeExpansionPending ||
+			resizeStatus == v1.PersistentVolumeClaimNodeExpansionInProgress ||
+			resizeStatus == v1.PersistentVolumeClaimNodeExpansionFailed {
+			if allocatedSize != nil {
+				newSize = *allocatedSize
+			}
+		}
+	} else {
+		// PV has already been expanded
+
+		// 1. If resizeStatus is in any of these states that means node is already working towards
+		// expansion and hence we don't need to perform an expanson
+		if resizeStatus == v1.PersistentVolumeClaimNodeExpansionInProgress ||
+			resizeStatus == v1.PersistentVolumeClaimNodeExpansionPending {
+			return resizeResponse
+		}
+
+		// 2. if resizeStatus is nil or any of partial completed states then, we should
+		// try and finish the expansion first
+		if pvc.Status.ResizeStatus == nil ||
+			resizeStatus == v1.PersistentVolumeClaimControllerExpansionInProgress ||
+			resizeStatus == v1.PersistentVolumeClaimControllerExpansionFailed ||
+			resizeStatus == v1.PersistentVolumeClaimNodeExpansionFailed {
+			if allocatedSize != nil {
+				newSize = *allocatedSize
+			}
+		}
```

Comment:
I perhaps miss something obvious, will this result in a new `ExpandVolumeDevice` call, even though e.g. Node expansion failed and there is nothing to do on the controller side?

Reply:
yeah - currently this "completes" the state-machine exhaustive condition checks. This can be optimized for intree controllers at least. But for exrernal-resizer - this is logic is bit more complicated because control-plane must reset the pvc status to `NodeExpansionPending` so as kubelet can safely retry.
![Screenshot from 2021-11-10 13-04-41](https://user-images.githubusercontent.com/278/141168612-5ac99ac4-595d-4b6e-8eed-1727c2e06df6.png)

Reply:
Fixed btw.